File: 001.Two Sum.py
class Solution(object):
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        for i in range(len(nums)):
            for j in range(len(nums) - i - 1):
                if (nums[i] + nums[(j + i + 1)]) == target:
                    return [i , j + i + 1]
        return []

class Solution(object):
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        maps = {}
        i = 0
        while i < len(nums):
            tmp = target - nums[i]
            if tmp in maps:
                return [maps[tmp], i]
            maps[nums[i]] = i
            i += 1
        return []

File: 001.Two Sum.swift
class Solution {
    // Approach #1 (Brute Force)
    func twoSum_1(_ nums: [Int], _ target: Int) -> [Int] {
        for i in 0..<nums.count {
            for j in i+1..<nums.count{
                if nums[i] + nums[j] == target {
                    return [i, j]
                }
            }
        }
        
        return []
    }
    
    // Approach #2 (Two-pass Hash Table)
    func twoSum_2(_ nums: [Int], _ target: Int) -> [Int] {
        var map = Dictionary<Int, Int>()
        
        for (index, num) in nums.enumerated() {
            map[num] = index
        }
   
        for (index, num) in nums.enumerated() {
            let complement = target - num
            if map[complement] != nil && map[complement] != index {
                    return [index, map[complement]!]
            }
        }
        
        return []
    }
    
    // Approach #3 (One-pass Hash Table)
    func twoSum_3(_ nums: [Int], _ target: Int) -> [Int] {
        var map = Dictionary<Int, Int>()

        for (index, num) in nums.enumerated() {
            let complement = target - num
            if map[complement] != nil && map[complement] != index {
                return [index, map[complement]!]
            }
            map[num] = index
        }
        
        return []
    }
}
